linkerd/http/upgrade/src/upgrade.rs

```diff
@@ -80,35 +78,40 @@ impl Http11Upgrade {
         Http11UpgradeHalves {
             server: Http11Upgrade {
                 half: Half::Server,
-                inner: inner.clone(),
+                inner: Some(inner.clone()),
             },
             client: Http11Upgrade {
                 half: Half::Client,
-                inner,
+                inner: Some(inner.clone()),
             },
         }
     }
 
     pub fn insert_half(self, upgrade: OnUpgrade) {
-        match self.half {
-            Half::Server => {
-                let mut lock = self
-                    .inner
+        match self {
+            Self {
+                inner: Some(inner),
+                half: Half::Server,
+            } => {
+                let mut lock = inner
                     .server
                     .try_lock()
                     .expect("only Half::Server touches server TryLock");
                 debug_assert!(lock.is_none());
                 *lock = Some(upgrade);
             }
-            Half::Client => {
-                let mut lock = self
-                    .inner
+            Self {
+                inner: Some(inner),
+                half: Half::Client,
+            } => {
+                let mut lock = inner
                     .client
                     .try_lock()
                     .expect("only Half::Client touches client TryLock");
                 debug_assert!(lock.is_none());
                 *lock = Some(upgrade);
             }
+            Self { inner: None, .. } => {}
```

Comment:
I don't think we should let this case occur gracefully -- if inset_half is called on a defunct clone of an upgrade request, it means the upgrade won't succeed, and we shouldn't continue as if it would. So insert_half should probably be converted to return a Result so that the [HTTP client](https://github.com/linkerd/linkerd2-proxy/blob/8b64841fc12c70f19a3276f6921b5aeaf3d046c0/linkerd/proxy/http/src/h1.rs#L139-L169) can bubble the error up to the caller.

Reply:
that's a great idea! 10f6c99 and 59d31d5 perform this transformation, propagating this error up in the HTTP/1 client.